### PR TITLE
Record key SCP metrics

### PR DESCRIPTION
**Pull request by marta-lokhova**
Record timings for the following parts of SCP:
* Time from nomination starting to first prepare 
* Time from first prepare to value externalizing

### CODE REVIEW

File: src/herder/HerderSCPDriver.cpp [hunk 1]
```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
 {
     mSCPMetrics.mAcceptedCommit.Mark();
 }
+
+void
+HerderSCPDriver::setMetricStartTime(uint64_t slotIndex,
+                                    VirtualClock::time_point start,
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: I would ditch the `start` parameter as it's always called with `now` (and rename the method to just be `recordSCPEvent`)

File: src/herder/HerderSCPDriver.cpp [hunk 2]
```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
+                                    bool isNomination)
+{
+    SCPTimingPtr timing;
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: don't use `shared_ptr` here, then you can simplify the code to just be:

`auto &timing = mSCPExecutionTimes[slotIndex];`

File: src/herder/HerderSCPDriver.cpp [hunk 3]
```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
+void
+HerderSCPDriver::recordSCPExecutionMetrics(uint64_t slotIndex)
+{
+    auto externalizeStart = mApp.getClock().now();
+
+    auto SCPTimingIt = mSCPExecutionTimes.find(slotIndex);
+    if (SCPTimingIt == mSCPExecutionTimes.end())
+    {
+        // No metric was recorded
+        CLOG(WARNING, "Herder")
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: remove logging as it's not useful

File: src/herder/HerderSCPDriver.cpp [hunk 4]
```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
+        {
+            // Update the metric
+            mSCPMetrics.mPrepareToExternalize.Update(prepareDelta);
+        }
+        CLOG(DEBUG, "Herder") << fmt::format("Prepare delta for slot {} is {}",
+                                             slotIndex, prepareDelta.count());
+    }
+
+    // Clean up timings map
+    mSCPExecutionTimes.erase(slotIndex);
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: you need to properly erase execution times, see how it's done with `mSCPTimers`

File: src/herder/HerderSCPDriver.cpp [hunk 5]
```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
+
+    auto checkThreshold = [&](std::chrono::nanoseconds delta) -> bool {
+        if (delta < Herder::NOMINATION_THRESHOLD_NANOSECS)
+        {
+            auto qset = mApp.getConfig().QUORUM_SET;
+
+            if (!(qset.innerSets.size() == 0 && qset.validators.size() == 1 &&
+                  qset.validators[0] == getSCP().getLocalNodeID()))
+            {
+                CLOG(WARNING, "Herder")
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: same here those logs are not useful

File: src/herder/HerderSCPDriver.cpp [hunk 6]
```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
+        // No metric was recorded
+        CLOG(WARNING, "Herder")
+            << "No SCP metrics recorded for slot, " << slotIndex
+            << " this should not have happened.";
+        return;
+    }
+
+    auto SCPTiming = SCPTimingIt->second;
+
+    auto checkThreshold = [&](std::chrono::nanoseconds delta) -> bool {
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: it seems that this lambda should instead take the parameters:
start, end, a reference to the timer to update and a string for the logging.

Then it's just a call like
`recordTiming(*SCPTiming->mNominationStart, *SCPTiming->mPrepareStart, mSCPMetrics.mNominateToPrepare, "nomination");`

File: src/herder/HerderSCPDriver.cpp [hunk 7]
```diff
@@ -722,4 +728,97 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
+        if (delta < Herder::NOMINATION_THRESHOLD_NANOSECS)
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: the code would be tighter if instead you were computing the threshold at the beginning of the function (outside of the lambda) as single node just means that the threshold is 0

File: src/herder/HerderSCPDriver.cpp [hunk 8]
```diff
@@ -53,6 +54,10 @@ HerderSCPDriver::SCPMetrics::SCPMetrics(Application& app)
           app.getMetrics().NewCounter({"herder", "state", "current"}))
     , mHerderStateChanges(
           app.getMetrics().NewTimer({"herder", "state", "changes"}))
+    , mNominateToPrepare(
+          app.getMetrics().NewTimer({"scp", "value", "nominate"}))
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: better names: `{"scp", "timing", "nominated" }` and `{"scp", "timing", "externalized"}`

File: src/herder/Herder.h [hunk 9]
```diff
@@ -55,6 +55,9 @@ class Herder
     // How many ledgers in the past we keep track of
     static uint32 const MAX_SLOTS_TO_REMEMBER;
 
+    // Min time between before first prepare (to avoid metric skewing)
+    static std::chrono::nanoseconds const NOMINATION_THRESHOLD_NANOSECS;
```
Comment (MonsieurNicolas) [inline src/herder/Herder.h]: comment doesn't really correspond to reality (it's not just about nomination):
this threshold is used to filter outliers when timing events.
I would just call it `TIMERS_THRESHOLD_NANOSEC`

File: src/herder/HerderSCPDriver.h [hunk 10]
```diff
@@ -164,11 +168,27 @@ class HerderSCPDriver : public SCPDriver
...
     };
 
     SCPMetrics mSCPMetrics;
 
+    struct SCPTiming
+    {
+        optional<VirtualClock::time_point> mNominationStart;
+        optional<VirtualClock::time_point> mPrepareStart;
+    };
+    using SCPTimingPtr = std::shared_ptr<SCPTiming>;
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.h]: per my earlier comment: ditch the shared_ptr

**Comment by marta-lokhova** [comment]
updated

File: src/herder/HerderSCPDriver.cpp [hunk 11]
```diff
@@ -730,95 +730,76 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
         return;
     }
 
     auto SCPTiming = SCPTimingIt->second;
 
-    auto checkThreshold = [&](std::chrono::nanoseconds delta) -> bool {
-        if (delta < Herder::NOMINATION_THRESHOLD_NANOSECS)
+    auto recordTiming = [&](VirtualClock::time_point start,
+                            VirtualClock::time_point end, medida::Timer& timer,
+                            std::string logStr) {
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: `logStr` should be const ref

File: src/herder/HerderSCPDriver.cpp [hunk 12]
```diff
@@ -730,95 +730,76 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
-    mSCPExecutionTimes[slotIndex] = timing;
 }
 
 void
 HerderSCPDriver::recordSCPExecutionMetrics(uint64_t slotIndex)
 {
     auto externalizeStart = mApp.getClock().now();
 
+    // Use threshold of 0 in case of a single node
+    auto qset = mApp.getConfig().QUORUM_SET;
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: use `auto& qset`

File: src/herder/HerderSCPDriver.cpp [hunk 13]
```diff
@@ -730,95 +730,76 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
-            << "No SCP metrics recorded for slot, " << slotIndex
-            << " this should not have happened.";
         return;
     }
 
     auto SCPTiming = SCPTimingIt->second;
 
-    auto checkThreshold = [&](std::chrono::nanoseconds delta) -> bool {
-        if (delta < Herder::NOMINATION_THRESHOLD_NANOSECS)
+    auto recordTiming = [&](VirtualClock::time_point start,
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: use `auto &SCPTiming`

**Comment by marta-lokhova** [comment]
fixed

File: src/herder/HerderSCPDriver.cpp [hunk 14]
```diff
@@ -722,4 +728,78 @@ HerderSCPDriver::acceptedCommit(uint64_t slotIndex, SCPBallot const& ballot)
...
+    if (SCPTimingIt == mSCPExecutionTimes.end())
+    {
+        return;
+    }
+
+    auto& SCPTiming = SCPTimingIt->second;
+
+    auto recordTiming = [&](VirtualClock::time_point start,
+                            VirtualClock::time_point end, medida::Timer& timer,
+                            const std::string& logStr) {
```
Comment (MonsieurNicolas) [inline src/herder/HerderSCPDriver.cpp]: we follow the convention `std::string const& `

**Comment by MonsieurNicolas** [comment]
r+ d27cb8ee995ff585829b750b2cf862b641510e5f

**Pull request merged**